### ISSUE TITLE
Feature Request - Cater for preferred versions constraints

### ISSUE THREAD

**Issue by andrewv99** [ddelange/pipgrip#24]
Hi,
pipgrip is a great library, but I can't see any feature to allow preferred versions to be specified. It is common that a particular version (or version range) of a library is preferred over others (Example reasons: compatibility with other software that uses it, free of particular bugs, etc.), Such a feature would be like specifying the dependencyManagament dependencies that Maven supports, where dependency version constraints can be specified. Allowing preferred versions to be specified independently to requirements.txt facilitates consistent version constraint specifications across multiple package builds.

**Possible Details**:
A requirements-constraints.txt file "dependency constrainer" could be optionally passed via the command line, which specifies the limited range or specific version(s) of packages that may be used - using the same version constraint syntax/semantics as requirements.txt. Let the requirements-constraints.txt dependency constraints be defined as X. Let the version constraints defined in requirements.txt be Y. The version constraint used by the dependency resolver for a given package Z from requirements.txt in Y would be:
- If only Y contains a constraint for Z, use the constraint in Y
- If Y and X contains a constraint for Z, use the intersection of Y and X. Raise an Exception if the intersection is empty.

Any plans to support such a feature (any time soon) ?

Thanks
Andrew

**Comment by ddelange** [comment]
Hi Andrew,

#### tl;dr
`pipgrip --install -r X -r Y` (and pass -v if you want to be warned about cyclic dependencies)

 So let's see if I get you right:

You're running pipgrip on your_package (let's say it depends on `requests` library, no version specified), and you have an environment in which their_package want to constrain e.g. `requests<1`. 

In this case, simply run `pipgrip --install your_package their_package` (this will do a `pip install their_package`, if it's already installed, pip won't touch it), or more specific `pipgrip --install your_package requests<1`. pipgrip will combine the trees of both and resolve the proper version ranges. This also works with pipgrip's -r flag for the constraints file you mentioned. -r can be passed multiple times if you have multiple files. If there's no viable combination of all the constraints in the scenario, you'll get an error with non-zero exit status. 

> If only Y contains a constraint for Z, use the constraint in Y

If you want Y to only 'constrain', and not 'require' (so Z not present in X -> do NOT install Z), you can use pip's builtin --constraint flag (-c, can also be used multiple times). pip constraint files (like pipgrip.lock) can only contain exact version pins. A pro however, is that they can contain packages that actually aren't needed in the install, you could even dump your whole environment into such a file (check out `pip freeze` output).

You can generate the lockfile for your_package and all your version ranges with `pipgrip --lock`, see also boto3 example in the README.

pipgrip's --install uses pip's --constraint in the background btw :) So in this case to generate (part of) the Y file:
- `pipgrip --lock your_package -r requirements-constraints.txt`
And then to install only the necessities (exactly) constrained in Y that are needed for your_package:
- `pip install your_package -c pipgrip.lock`

Does that solve your issue?

**Comment by andrewv99** [comment]
Hi @ddelange 

Many thanks for the detailed response :)
Yes, that resolves the dependency issue.

Thanks
Andrew
